drivers/slipmux: Add to RIOT

Teufelchen1 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

Hey 🪼,

this adds an (incomplete) implementation of [Slipmux](https://tools.ietf.org/html/draft-bormann-t2trg-slipmux-03) to RIOT. It extends the current slip driver and is backwards compatible.

The second commit adds an usage example, where the shell commands are exposed to coap and through it to slipmux. That commit will be dropped prio merging. 

It is incomplete as frame abort is missing. 

### Testing procedure

Grab your favorite slipmux tool and connect RIOTs stdio to it. Make sure to build RIOT with `USEMODULE += slipdev_stdio` and optionally set `CFLAGS += "-DCONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX=10"` to be able to handle coap packets that are longer than 64 bytes. See the changes I made to the default example. 


### Feedback wanted: coap server

My current approach spawns a new thread with a nanocoap server to handle the incoming and outgoing configuration frames. This might cause issues when another coap server is in use (one that is connected to the network instead of just the uart). For example, the `NANOCOAP_RESOURCE(..)`s will be shared between them.

I also experimented with sending the coap messages with a fake ip6+udp header into gnrc. One can then only run one single coap server (and it is no longer bound to nanocoap) who listens for such packets. The outgoing way is a bit more annoying, here a new dummy interface must be created, to which gnrc can direct those fake ip6+udp headers. Very similar to how it is already done with the current slip implementation.

What alternative approaches do you see? Any idea is appreciated.

Teufelchen1 commented:
From the byte-wise overhead, it is not that much of a difference yes. I don't think it is that much of a hassle tho. Beside, as soon as you use IP+UDP you have endpoints in a network and those need to be configured (I need to know the IP I want to talk to). Doing slipmux configuration frames you have a guaranteed point to point channel with no configuration nor hassle ;)

chrysn commented:
I can't comment on the code right now, but maybe for the open aspects:

* "Why?" Teufelchen already mentioned the configuration aspect; without precise knowledge of the concrete application (and a point of using SLIP is that this knowledge is not required) the host may not even know that it reaches the board this way. In particular,
  - this connection may be more trusted than a network connection: While I generally advocate no-TOFU / all-links-are-untrusted / cryptographically-secure-or-bust policy, TOFU policies do have their places, and TOFU over a link that is known not to be networked is preferable over a link where one would need to apply heuristics to know whether a peer is local. (Like, sure, might do link-local-only, but maybe there's NDP-proxying involved?).
    - This might be a good fit also for devices that don't do network at all, or not on that interface. On those, actual SLIP packets would be blackhole'd, but CoAP provides an interface that is easier to automate than shell interactions, and moreover easily ported to devices where no serial line but network is available.
    -  Where we do have Ethernet-over-USB, I think that a dedicated CoAP-over-USB protocol would be preferred. However, I'm not aware of any such attempt, and even then, there are relatively new devices around still that still run through the moral equivalent of an FTDI chip.

* "Which CoAP server to run?" I think we should run the same single CoAP server on all transports as a matter of simplicity; where a UART connection is more trusted than a network peer, which transport it is should be discoverable by inspecting the socket-ish thing it comes in over. I don't have an opinion yet on which implementation strategy I'd advocate for.

miri64 commented on drivers/include/slipdev.h:
```diff
@@ -93,6 +93,14 @@ enum {
      */
     SLIPDEV_STATE_STDIN_ESC,
+    /**
+     * @brief   Device writes received data as coap message
+     */
+    SLIPDEV_STATE_CONFIG,
+    /**
+     * @brief   Device writes received data as coap message, next byte is escaped
```
```suggestion
     * @brief   Device writes received data as CoAP message
     */
    SLIPDEV_STATE_CONFIG,
    /**
     * @brief   Device writes received data as CoAP message, next byte is escaped
```

miri64 commented on drivers/include/slipdev.h:
```diff
@@ -120,13 +128,19 @@ typedef struct {
+#if IS_USED(MODULE_SLIPDEV_STDIO)
+    chunk_ringbuf_t rb_config;              /**< Ringbuffer to store received CONFIG frames.*/
+    uint8_t rxmem_config[CONFIG_SLIPDEV_BUFSIZE]; /**< memory used by RX buffer */
+    kernel_pid_t coap_server_pid;
```
```suggestion
    kernel_pid_t coap_server_pid;           /**< The PID of the CoAP server */
```

Wrt to future extensions, maybe that can be more generalized so that, e.g., also a gcoap or unicoap queue could be put here, if they exist to not have to provide an extra server then.

Again, mainly a comment for future improvements, this does not have to be addressed in this PR.

benpicco replied:
Please don't foce `slipdev_stdio` to require a CoAP server, make this a separate (pseudo)module.

miri64 commented on drivers/slipdev/include/slipdev_internal.h:
```diff
@@ -43,11 +43,51 @@ extern "C" {
 #define SLIPDEV_STDIO_START (0x0aU)
+
+/**
+ * @brief   Marker byte for beginning of configuration/coap
```
```suggestion
 * @brief   Marker byte for beginning of configuration/CoAP
```

miri64 commented on drivers/slipdev/include/slipdev_internal.h:
```diff
@@ -43,11 +43,51 @@ extern "C" {
+#define SLIPDEV_CONFIG_START (0xa9U)
 /** @} */
 
+
```
`uncrustify` is complaining about too many consecutive lines here.

```suggestion
```

miri64 commented on drivers/slipdev/include/slipdev_internal.h:
```diff
@@ -43,11 +43,51 @@ extern "C" {
+
+/* Special init
+ * the result of normal fcs init combined with slipmux config start (0xa9)
+ */
+#define SPECIAL_INIT_FCS (0x374cU)
+#define INIT_FCS (0xffffU)
+#define GOOD_FCS (0xf0b8U)
+
+static const uint16_t FCS_LOOKUP[256] = {
```
Not sure why Doxygen is complaining here. Maybe it helps to already put them in a Doxygen group?

miri64 commented on drivers/slipdev/include/slipdev_internal.h:
```diff
@@ -43,11 +43,51 @@ extern "C" {
+
+/* Special init
+ * the result of normal fcs init combined with slipmux config start (0xa9)
+ */
+#define SPECIAL_INIT_FCS (0x374cU)
+#define INIT_FCS (0xffffU)
+#define GOOD_FCS (0xf0b8U)
+
+static const uint16_t FCS_LOOKUP[256] = {
```
Does the `static const` lookup table really need to be in a header? This could potentially duplicate it everytime the header is included into a C file.

I'd rather have the FCS checking API you already have in https://github.com/RIOT-OS/RIOT/pull/21418/files#diff-b7bda8876512e2948e36a07d7926bd6d15c4b7bb1f043e832daaae113aadbb32R411 ff public and all the constants here private.

Teufelchen1 replied:
I moved it.

miri64 commented on drivers/slipdev/slipdev.c:
```diff
@@ -31,8 +32,11 @@
+#include "net/nanocoap.h"
 #include "stdio_uart.h"
 
+static char coap_stack[1024];
```
```suggestion
#define COAP_STACKSIZE (1024)
/* but better make this a public, configurable macro CONFIG_SLIPDEV_COAP_STACKSIZE? */

static char _coap_stack[COAP_STACKSIZE];
```

miri64 commented on drivers/slipdev/slipdev.c:
```diff
@@ -80,6 +84,45 @@ static void _slip_rx_cb(void *arg, uint8_t byte)
+        default:
+            /* discard frame if byte can't be added */
+            if (!crb_add_byte(&dev->rb_config, byte)) {
+                DEBUG("slipdev: rx buffer full, drop frame\n");
```
I see that this sneaked in below as well but it is not a good idea to try to print something in

1. ISR context
2. A UART handler callback

miri64 commented on drivers/slipdev/stdio.c:
```diff
@@ -34,9 +34,11 @@ static void _isrpipe_write(void *arg, uint8_t data)
+     * the network device is initialized */
     uart_init(slipdev_params[0].uart, slipdev_params[0].baudrate,
               _isrpipe_write, &stdin_isrpipe);
+    
```
Trailing whitespaces (Github somehow shows that as a complete line removal, but I just want you to remove the non-linebreaking whitespaces)

```suggestion

```

miri64 commented on examples/basic/default/Makefile:
```diff
@@ -27,7 +27,10 @@ DEVELHELP ?= 1
 # Change this to 0 show compiler invocation lines by default:
 QUIET ?= 1
 
+CFLAGS += "-DCONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX=10"
```
Does this need to be this block size? If yes, please make it configurable via Kconfig (`app.config`) so users do not run into the `CFLAGS` being defined twice.

miri64 commented on examples/basic/default/Makefile:
```diff
@@ -27,7 +27,10 @@ DEVELHELP ?= 1
+CFLAGS += "-DCONFIG_NANOCOAP_BLOCK_SIZE_EXP_MAX=10"
+
 # Modules to include:
+USEMODULE += slipdev_stdio
```
I would prefer this to be its own example.

Teufelchen1 replied:
> The second commit adds an usage example, where the shell commands are exposed to coap and through it to slipmux. That commit will be dropped prio merging.

Thanks for the review!

Teufelchen1 replied:
I resolved your other comments without fixing them as they would be dropped too.

miri64 commented on sys/include/shell.h:
````diff
@@ -301,6 +301,17 @@ int shell_parse_file(const shell_command_t *commands,
  * ```
  */
+
+#include "net/nanocoap.h"
+__attribute__ ((unused))
+static ssize_t _cmd_handler(coap_pkt_t *pkt, uint8_t *buf, size_t len, coap_request_ctx_t *context)
+{
+
+    char * desc = (char *) context->resource->context;
+    return coap_reply_simple(pkt, COAP_CODE_205, buf, len,
+            COAP_FORMAT_TEXT, (uint8_t*)desc, strlen(desc));
+}
````
I'd expect some guard here for the inclusion of the `slipmux` module.

miri64 commented on sys/include/shell.h:
```diff
@@ -309,6 +320,10 @@ int shell_parse_file(const shell_command_t *commands,
         .desc = _xfa_ ## cmd ## _cmd_desc, \
         .handler = &func \
+    }; \
+    extern ssize_t _cmd_handler(coap_pkt_t *pkt, uint8_t *buf, size_t len, coap_request_ctx_t *context); \
+    NANOCOAP_RESOURCE(cmd) { \
+        .path = "/shell/" #cmd, .methods = COAP_GET, .handler = _cmd_handler, .context = help \
```
I'd expect some guard here for the inclusion of the slipmux module.

miri64 commented on sys/include/shell.h:
```diff
@@ -309,6 +320,10 @@ int shell_parse_file(const shell_command_t *commands,
         .desc = _xfa_ ## cmd ## _cmd_desc, \
         .handler = &func \
+    }; \
+    extern ssize_t _cmd_handler(coap_pkt_t *pkt, uint8_t *buf, size_t len, coap_request_ctx_t *context); \
```
Line too long
```suggestion
    extern ssize_t _cmd_handler(
        coap_pkt_t *pkt, uint8_t *buf, size_t len, coap_request_ctx_t *context); \
```

miri64 reviewed (changes requested):
First round of review.

benpicco commented on drivers/slipdev/slipdev.c:
```diff
@@ -355,8 +440,82 @@ static const netdev_driver_t slip_driver = {
+
 void slipdev_setup(slipdev_t *dev, const slipdev_params_t *params, uint8_t index)
 {
+#if IS_USED(MODULE_SLIPDEV_STDIO)
```
Please use a separate pseudo-module for this, normal slipdev stdio multiplex should not require a CoAP server.

benpicco replied:
If you desire to do so go ahead.
I just want to avoid bloating the slipdev driver for simple device to device links that might have a multiplexed debug interface for convenience.

Teufelchen1 replied:
That will be much more work than I thought. I'll move that task to a followup PR. For this one, I adopted your wish and guarded it with its own pseudomodule.
In addition, I changed the way the parsing is guarded with the `#if IS_USED(..)`. Before: The whole case of stdio was removed if the module was not selected. Now: Only the data processing is removed, the cases for parsing remain. (Same for configuration frames). Reason is simple: If you remove the parsing + states, any byte received will be treated as a IP packet and then dropped by gnrc as "undefined packet type". This no-longer the case, instead, receiving a diagnostic/configuration frame without their pseudomodules enabled, they will be silently ignored.

benpicco commented on drivers/slipdev/slipdev.c:
```diff
@@ -355,8 +440,82 @@ static const netdev_driver_t slip_driver = {
 };
 
+#if IS_USED(MODULE_SLIPDEV_STDIO)
+static uint16_t _fcs_part(const uint8_t * data, size_t len) {
```
The RFC says

> After the CoAP message, the last two bytes of a CoAP frame therefore contain a 16-bit CRC FCS computed over the byte 0xA9 followed by the (unescaped) bytes of the CoAP message, computed as specified in [[RFC1662](https://datatracker.ietf.org/doc/html/rfc1662)].

benpicco commented on drivers/slipdev/slipdev.c:
```diff
@@ -355,8 +440,82 @@ static const netdev_driver_t slip_driver = {
+    uint16_t fcs = SPECIAL_INIT_FCS;
+    for (size_t i = 0; i < len; ++i)
+    {
+        fcs = (fcs >> 8) ^ FCS_LOOKUP[(fcs ^ data[i]) & 0xff];
```
This looks a lot like `crc16_ccitt_kermit_update()`

Teufelchen1 replied:
You know whats funny? I wrote the unittest for that function.

Teufelchen1 commented:
@benpicco I think this is ready for another review. If that is positive, I will drop the shell/example related commit and squash the others.

miri64 commented on drivers/include/slipdev.h:
```diff
@@ -120,13 +129,19 @@ typedef struct {
     netdev_t netdev;                        /**< parent class */
     slipdev_params_t config;                /**< configuration parameters */
-    chunk_ringbuf_t rb;                     /**< Ringbuffer to store received frames.       */
+    chunk_ringbuf_t rb;                     /**< Ringbuffer to store received NET frames.   */
```
Is “NET frames” wording from the draft?

Teufelchen1 replied:
no it is not, it would be "Packet Transfer". I thought I would assume a bit that the reader is more likely to know RIOT than the Draft, so I picked a name that represents "Packet Transfer" for RIOT networking humans. Do you think that was a poor choice? I have no hard feelings on this. ("Packet frames" is rather confusing as one could say "coap packet")

Teufelchen1 replied:
ah yes, I can see that as confusing. I applied your suggestion.

miri64 commented on drivers/include/slipdev.h:
```diff
@@ -120,13 +129,19 @@ typedef struct {
     uint8_t rxmem[CONFIG_SLIPDEV_BUFSIZE];  /**< memory used by RX buffer */
+
+#if IS_USED(MODULE_SLIPDEV_CONFIG)
+    chunk_ringbuf_t rb_config;              /**< Ringbuffer to store received CONFIG frames.*/
```
Same here for “CONFIG frames”.

mguetschow commented on drivers/slipdev/slipdev.c:
```diff
@@ -348,15 +424,70 @@ static const netdev_driver_t slip_driver = {
+            crb_consume_chunk(&dev->rb_config, buf, len);
+
+            /* Is crc correct via Residue test */
+            if (crc16_ccitt_fcs_update(SPECIAL_INIT_FCS, buf, len) != 0xF0B8) {
```
where does this constant come from? mind using a define or at least add a comment?

Teufelchen1 replied:
That's a residue test, the magic value is defined [here, in the crc header](https://github.com/RIOT-OS/RIOT/blob/3f498686cf074f980480f7c6117ee00b55cd323e/sys/include/checksum/crc16_ccitt.h#L130). Please don't make me add all crcs residue values as defines :fearful:

mguetschow replied:
Hum, while that would be way cleaner IMO, I think it is out of scope for this PR.

mguetschow commented:
Some high-level comments from my side:

- incomplete implementation

> this adds an (incomplete) implementation of [Slipmux](https://tools.ietf.org/html/draft-bormann-t2trg-slipmux-03) to RIOT. It extends the current slip driver and is backwards compatible. (...) It is incomplete as frame abort is missing.

Isn't it also missing [diagnostic transfer](https://datatracker.ietf.org/doc/html/draft-bormann-t2trg-slipmux-03#section-4)? Just had a quick glance at the code, but could only find references to configuration frames.

- documentation: before merging this should definitely get documentation, probably on the defined pseudomodule `slipdev_config`. Not sure if this may even warrant its own `doc.md` somewhere? It should definitely mention `slipmux` and clearly list what is currently supported. Also some notes on the implementation (currently spawning an additional thread, sharing `NANOCOAP_RESOURCE` with any other nanocoap server) should be in the documentation.

- coap server with different backends: I would hope unicoap could help with that. I'm pretty sure it supports defining resources only for a subset of driver backends (e.g. just slipmux and not udp), which means that a single unicoap server could be used for several backends. Maybe @carl-tud has some more comments on this?

- nanocoap/nanocoap_sock: could you maybe explain in a bit more detail how this actually works right now? I see the "server thread" just receives chunks from the ringbuffer, tries to parse and answer them (using nanocoap). Every chunk is supposed to be a CoAP packet as received within slipdev, right? Where do the `NANOCOAP_RESOURCE`s (which are part of `nanocoap_sock` afaict) actually come into play?

Teufelchen1 commented:
> Just trying to safeguard against follow-up work being stalled at some point, where we would end up with yet another feature not documented at all.

Yes, that's your job :p Speaking of non documentation, I am either overlooking something or slip has already been poorly documented :/ Either way, I roughly agree with you but there are remarks.

> [Add] very short documentation for now (probably in the pseudomodule file then) just mentioning:

I opted against that and followed the example of the already existing documentation for slipdev_stdio.

>    this is slipmux configuration with a link to the draft

Yes, I added the links. Fun fact, I also fixed the broken link of SLIP dev itself :D

>    @warning stating this is WIP / experimental work

Not sure tbh. So far, it not claimed anywhere this is "rfc conform", as such, it can not be incomplete. It might be considered WIP? But what would that even mean for a pseudomodule that already works exactly like one would imagine? (still, you could persuade me to add that)

>    briefly list the current shortcomings you mention above (or, alternatively link to this PR)

There are no shortcomings for the user as "slipmux conform" is not claimed. Things would look different if one would enable a `slipmux` module but you don't. e.g. If one just uses slip (without stdio/configuration), the code is conform to SLIP rfc but not to Slipmux draft. This mess is part of the idea I outlined a bit before:

In future work, one will have to flip it on it's head. There will be one module called 'slipmux' that will enable all three frame types and be slipmux-draft-conform. There will be three sub-modules each guarding one of the three frame types, enabling you to only have partial slipmux. There will be a module called `slip` that is a alias for the submodule that enables the packet transfer and that will be slip-rfc conform.
(Just how I imagine it, nothing concrete yet)

Teufelchen1 commented on drivers/include/slipdev.h:
```diff
@@ -23,11 +26,28 @@
  * @ingroup     drivers_netdev
  * @brief       SLIP network device over @ref drivers_periph_uart
- * @see         [RFC 1055](https://github.com/RIOT-OS/RIOT/pull/6487)
+ * @see         [RFC 1055](https://datatracker.ietf.org/doc/html/rfc10557)
```
nice catch!

miri64 reviewed (commented):
Did not test this and did not look into the whole state machine, but I trust @benpicco's review on that (though please reply if @Teufelchen1 answers were to your liking). So from my side, we can go, if @mguetschow does not find anything else :-)

carl-tud commented:
@mguetschow 
> * coap server with different backends: I would hope unicoap could help with that. I'm pretty sure it supports defining resources only for a subset of driver backends (e.g. just slipmux and not udp), which means that a single unicoap server could be used for several backends. Maybe @carl-tud has some more comments on this?

a, The notion of a CoAP server is loosely defined. A driver can open any number of sockets or for that matter 'concepts' similar to sockets. Each driver then forwards all packets to 'the unicoap server' after having removed the outer frame (UDP headers, etc). Information on what transport has been used is retained and optionally used to restrict resources to certain transports. I you really want to, you _can_ handle this manually. There is an API for that.

b, Is this what Lasse wants: multiple ports and thus multiple sockets? I yes, I talked to Lasse off channel and we agreed on solution I have yet to implement. But, adding this to `unicoap` is almost trivial.

Teufelchen1 commented:
@miri64 you are still requesting change

mguetschow commented on drivers/slipdev/slipdev.c:
```diff
@@ -348,15 +426,69 @@ static const netdev_driver_t slip_driver = {
+    while (1) {
+        thread_flags_wait_any(1);
+        size_t len;
+        while (crb_get_chunk_size(&dev->rb_config, &len)) {
+            crb_consume_chunk(&dev->rb_config, buf, len);
```
fun :eyes:
